clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -237,6 +245,72 @@ class BuiltinNameEmitter {
   // same entry (<I1, I2, I3>).
   MapVector<BuiltinIndexListTy *, BuiltinTableEntries> SignatureListMap;
 };
+
+/// Generate a header file representing the builtins as understood by clang.
+/// This allow the generation of a library interface.
+class JSONBuiltinInterfaceEmitter : public BuiltinNameEmitter {
+public:
+  struct TypeDesc {
+    TypeDesc(const Record *T);
+    // Return a string representing the "base" type (ignores pointers).
+    std::string GetBaseTypeAsStr() const;
+
+    const Record *Rec;
+    llvm::StringRef Name;
+    bool IsOpaque;
+    // Size of the vector (if applicable).
+    int VecWidth;
+    // Size of the element in bits.
+    int Size;
+    // Is a integer.
+    bool IsInteger;
```

Comment:
It looks like a lot of these are mutually exclusive, or at least closely related.  Consider using enums/masks instead of bools for a lot of this.

particularly things like CV qualification.

Reply:
This maps the fields as defined in td fields. But I don't see what we would gain, this is exclusively used in the file and it is intended to process 1 file in the whole code base.

Reply:
The hope is that it would be more readable as well as more consistent with the rest of clang code.

Reply:
Moved the complexity from one place to another.